src/core/OSTreeTUI.hpp

```diff
@@ -3,20 +3,139 @@
+#include "ftxui/component/screen_interactive.hpp"  // for ScreenInteractive
+#include "ftxui/dom/elements.hpp"  // for Element, operator|, text, center, border
+
+#include "commit.hpp"
+#include "footer.hpp"
+#include "manager.hpp"
+
 #include "../util/cpplibostree.hpp"
 
-namespace OSTreeTUI {
+class OSTreeTUI {
+public:
+    /**
+     * @brief Constructs, builds and assembles all components of the OSTreeTUI.
+     * 
+     * @param repo ostree repository (OSTreeRepo)
+     * @param startupBranches optional list of branches to pre-select at startup (providing nothing will display all branches)
+     */ 
+    explicit OSTreeTUI (const std::string& repo, const std::vector<std::string> startupBranches = {});
+    
+    /**
+     * @brief Runs the OSTreeTUI (starts the ftxui screen loop).
+     * 
+     * @return exit code
+     */ 
+    int run();
+
+    /// @brief Refresh Level 3: Refreshes the commit components
+    void refresh_commitComponents();
+
+    /// @brief Refresh Level 2: Refreshes the commit list component & upper levels
+    void refresh_commitListComoponent();
+
+    /// @brief Refresh Level 1: Refreshes complete repository & upper levels
+    bool refresh_repository();
 
     /**
-     * @brief OSTree TUI main
+     * @brief sets the promotion mode
+     * @param active activate (true), or deactivate (false) promotion mode
+     * @param hash provide if setting mode to true
+     * @return false, if other promotion gets overwritten
+     */
+    bool setPromotionMode(bool active, std::string hash = "", bool setPromotionBranch = true);
+
+    /** @brief promote a commit
+     * @param hash 
+     * @param branch 
+     * @return promotion success
+     */ 
+    bool promoteCommit(std::string hash, std::string branch, std::vector<std::string> metadataStrings = {}, std::string newSubject = "", bool keepMetadata = true);
+
+private:
+    /**
+     * @brief Calculates all visible commits from an OSTreeRepo and a list of branches.
      * 
-     * @param repo ostree repository path
+     * @param repo OSTreeRepo
+     * @param visibleBranches Map: branch name -> visible
+     * @return Complete list of commit hashes in repo, that are part of the given branches 
+     */
+    std::vector<std::string> parseVisibleCommitMap(cpplibostree::OSTreeRepo& repo, std::unordered_map<std::string, bool>& visibleBranches);
+
+    /**
+     * @brief Adjust scroll offset to fit the selected commit
      */
-    int main(const std::string& repo, const std::vector<std::string>& startupBranches = {}, bool showTooltips = true);
+    void adjustScrollToSelectedCommit();
+
+public:
+    // SETTER
+    void setPromotionBranch(std::string promotionBranch);
+    void setSelectedCommit(size_t selectedCommit);
+
+    // non-const GETTER
+    std::vector<std::string>& getColumnToBranchMap();
+    ftxui::ScreenInteractive& getScreen();
+    
+    // GETTER
+    const cpplibostree::OSTreeRepo& getOstreeRepo() const;
```

Comment:
```suggestion
    [[nodiscard]] const cpplibostree::OSTreeRepo& getOstreeRepo() const;
```

Please add `[[nodiscard]]` where ever possible especially for all your `const` getters. This is just a way of saying hey if you invoke this method you have to deal with the return value. You can not "discard" it. If you discard it, it's most likely a bug. Ref: https://en.cppreference.com/w/cpp/language/attributes/nodiscard

Reply:
Oh, interesting, didn't know about that attribute. Will add it where ever it makes sense!